tools/adios2pio-nm/adios2pio-nm-lib.cxx

```diff
@@ -1525,229 +1569,337 @@ int ConvertVariableTimedPutVar(IOVector &bpIO, EngineVector &bpReader,
+    assert(elemsize > 0);
+
     /* Calculate how many records/steps we have for this variable */
-    int nsteps = 1;
-    int ts = 0; /* Loop from ts to nsteps, below ts may become the last step */
+    /* Compute the number of time steps */
+    int nsteps = 0;
+    int32_t *decomp_buffer = NULL;
+    int32_t *frame_buffer = NULL;
+    char *fillval_buffer = NULL;
+    int fillval_idx = 0;
 
-    /* Compute the total number of blocks */
-    int l_nblocks = 0;
-    int g_nblocks = 0;
-    for (size_t i = 1; i <= wfiles.size(); i++)
+    try
     {
-        try
+        /* get number of application steps in adios step */
+        adios2::Variable<int> frameid_var = bpIO[0].InquireVariable<int>("/__pio__/track/frame_id/" + variable_name);
+        if (!frameid_var)
         {
-            *v_base = bpIO[i].InquireVariable<T>(varname);
-            const auto vb_blocks = bpReader[i].BlocksInfo(*v_base, 0);
-            l_nblocks += vb_blocks.size();
+            return BP2PIO_ERROR;
         }
-        catch (const std::exception &e)
+        const auto vb_blocks = bpReader[0].BlocksInfo(frameid_var, time_step);
+        for (size_t i = 0; i < vb_blocks.size(); i++)
         {
-            ierr = BP2PIO_ERROR;
-            break;
+            nsteps += vb_blocks[i].Count[0];
         }
-        catch (...)
+
+        /* read frame_id buffer */
+        frame_buffer = (int32_t*)calloc(nsteps, sizeof(int32_t));
+        int tmp_idx = 0;
+        for (size_t i = 0; i < vb_blocks.size(); i++)
         {
-            ierr = BP2PIO_ERROR;
-            break;
+            frameid_var.SetBlockSelection(i);
+            bpReader[0].Get(frameid_var, frame_buffer+tmp_idx, adios2::Mode::Sync);
+            tmp_idx += vb_blocks[i].Count[0];
         }
-    }
-    ERROR_CHECK_RETURN(ierr, err_val, err_cnt, comm)
-
-    MPI_Allreduce(&l_nblocks, &g_nblocks, 1, MPI_INT, MPI_SUM, comm);
 
-    nsteps = g_nblocks / nblocks_per_step;
-    if (g_nblocks != nsteps * nblocks_per_step)
-    {
-        if (debug_out)
-            cout << "rank " << mpirank << ":ERROR in processing darray '" << varname
-                 << "'. Number of blocks = " << g_nblocks
-                 << " does not equal the number of steps * number of writers = "
-                 << nsteps << " * " << nblocks_per_step << " = " << nsteps*nblocks_per_step
-                 << endl;
-    }
+        /* read decomp_id buffer */
+        adios2::Variable<int> decompid_var = bpIO[0].InquireVariable<int>("/__pio__/track/decomp_id/" + variable_name);
+        if (!decompid_var)
+        {
+            return BP2PIO_ERROR;
+        }
+        const auto db_blocks = bpReader[0].BlocksInfo(decompid_var, time_step);
+        decomp_buffer = (int32_t*)calloc(nsteps, sizeof(int32_t));
+        tmp_idx = 0;
+        for (size_t i = 0; i < db_blocks.size(); i++)
+        {
+            decompid_var.SetBlockSelection(i);
+            bpReader[0].Get(decompid_var, decomp_buffer + tmp_idx, adios2::Mode::Sync);
```

Comment:
I am assuming SetBlockSelection() & Get() failure throws an exception

Reply:
Yes, it is called inside a try block.